tests/devices/qubit/test_sampling.py

```diff
@@ -47,80 +48,207 @@ def samples_to_probs(samples, num_wires):
+        state[:, :, 1] = 0.5  # third wire is always 1
+        alltrue_axis = wire_order.index(2)
+
+        samples = sample_state(state, 20, wires=wire_order)
+        assert all(samples[:, alltrue_axis])
+
+    def test_sample_state_custom_rng(self):
+        """Tests that a custom RNG can be used with sample_state."""
+        custom_rng = np.random.default_rng(12345)
+        samples = sample_state(two_qubit_state, 4, rng=custom_rng)
+        expected = [[0, 1], [0, 1], [1, 0], [1, 0]]
+        assert qml.math.allequal(samples, expected)
+
+    def test_approximate_probs_from_samples(self, init_state):
+        """Tests that the generated samples are approximately as expected."""
+        n = 4
+        shots = 20000
+        state = init_state(n)
+
+        flat_state = state.flatten()
+        expected_probs = np.real(flat_state) ** 2 + np.imag(flat_state) ** 2
+
+        samples = sample_state(state, shots)
+        approx_probs = samples_to_probs(samples, n)
+        assert np.allclose(approx_probs, expected_probs, atol=APPROX_ATOL)
+
+    def test_entangled_qubit_samples_always_match(self):
+        """Tests that entangled qubits are always in the same state."""
+        bell_state = np.array([[1, 0], [0, 1]]) / np.sqrt(2)
+        samples = sample_state(bell_state, 1000)
+        assert samples.shape == (1000, 2)
+        assert not any(samples[:, 0] ^ samples[:, 1])  # all samples are entangled
+        assert not all(samples[:, 0])  # some samples are |00>
+        assert any(samples[:, 0])  # ...and some are |11>!
+
+    @pytest.mark.slow
+    @pytest.mark.parametrize("num_wires", [13, 14, 15, 16])
+    def test_sample_state_many_wires(self, num_wires):
+        """Tests that sample_state works as expected with many wires, and with re-ordering."""
+        shots = 10000
+        shape = (2,) * num_wires
+        flat_state = np.arange(1, 2**num_wires + 1, dtype=np.float64)
+        original_norm = np.linalg.norm(flat_state)
+        flat_state /= original_norm
+        state = flat_state.reshape(shape)
+        expected_probs = np.real(flat_state) ** 2 + np.imag(flat_state) ** 2
+
+        ordered_samples = sample_state(state, shots)
+        ordered_probs = samples_to_probs(ordered_samples, num_wires)
+        assert np.allclose(ordered_probs, expected_probs, atol=APPROX_ATOL)
+
+        random_wires = list(range(num_wires))
+        shuffle(random_wires)
+        random_samples = sample_state(state, shots, wires=random_wires)
+        random_probs = samples_to_probs(random_samples, num_wires)
+
+        reordered_probs = ordered_probs.reshape(shape).transpose(random_wires).flatten()
+        assert np.allclose(reordered_probs, random_probs, atol=APPROX_ATOL)
+
+
+class TestMeasureSamples:
+    """Test that the measure_with_samples function works as expected"""
+
+    def test_sample_measure(self):
+        """Test that a sample measurement works as expected"""
+        state = qml.math.array(two_qubit_state)
+        shots = 100
```

Comment:
```suggestion
        shots = qml.measurements.Shots(100)
```

Since in the future, `measure_with_samples` will need to also handle shot vectors, it needs to accept a `Shots` object instead of just an `int`.  Duplicate this comment to all the below tests.